src/JMS/Serializer/GraphNavigator.php

```diff
@@ -108,6 +108,21 @@ public function accept($data, array $type = null, Context $context)
             $type = array('name' => 'NULL', 'params' => array());
         }
 
+        // Trigger pre-serialization callbacks, and listeners if they exist.
+        // Dispatch pre-serialization event before handling data to have ability change type in listener
+        if ($context instanceof SerializationContext) {
+            if (null !== $this->dispatcher && $this->dispatcher->hasListeners('serializer.pre_serialize', $type['name'], $context->getFormat())) {
+                $this->dispatcher->dispatch('serializer.pre_serialize', $type['name'], $context->getFormat(), $event = new PreSerializeEvent($context, $data, $type));
+                $type = $event->getType();
+            }
+        } elseif ($context instanceof DeserializationContext) {
+            if (null !== $this->dispatcher && $this->dispatcher->hasListeners('serializer.pre_deserialize', $type['name'], $context->getFormat())) {
+                $this->dispatcher->dispatch('serializer.pre_deserialize', $type['name'], $context->getFormat(), $event = new PreDeserializeEvent($context, $data, $type));
+                $type = $event->getType();
+                $data = $event->getData();
+            }
+        }
```

Comment:
your change is wrong. You are now dispatching the event before the level handling in the context, which does not work properly anymore